src/snowflake/snowpark/_internal/analyzer/metadata_utils.py

```diff
@@ -2,51 +2,156 @@
 # Copyright (c) 2012-2024 Snowflake Computing Inc. All rights reserved.
 #
 
-from typing import List, Optional
+from typing import TYPE_CHECKING, DefaultDict, Dict, List, Optional
 
-from snowflake.snowpark._internal.analyzer.expression import Attribute
+from snowflake.snowpark._internal.analyzer.expression import Attribute, Expression, Star
 from snowflake.snowpark._internal.analyzer.snowflake_plan_node import Limit, LogicalPlan
+from snowflake.snowpark._internal.analyzer.unary_expression import UnresolvedAlias
 
+if TYPE_CHECKING:
+    from snowflake.snowpark._internal.analyzer.analyzer import Analyzer
 
-def infer_metadata(
-    source_plan: LogicalPlan,
-) -> Optional[List[Attribute]]:
+
+class PlanMetadata:
     """
-    Infer metadata from the source plan.
-    Returns the metadata including attributes (schema).
+    Metadata of a plan including attributes (schema) and quoted identifiers (column names).
```

Comment:
Oh sorry, i don't really mean to wrap the init with just source plan. I kind of want to make this class similar as the internalFrame in our pandas that is immutable, every time there is change just create a new one instead of do inplace update
```
@dataclass(frozen=True)
class InternalFrame:
```

The infer_quoted_identifiers_from_expressions and infer_medata data can still work like before be provided as utility function.

Reply:
no worries, done